unpacker.go

```diff
@@ -0,0 +1,338 @@
+				}
+			}
+
+			chain = append(chain, states[i].layer.Diff.Digest)
+			states[i].unpacked = true
+			log.G(ctx).WithFields(logrus.Fields{
+				"desc": states[i].layer.Blob,
+				"diff": states[i].layer.Diff,
+			}).Debug("layer unpacked")
+		}
+	}
+
+	for _, state := range states {
+		if !state.unpacked && !state.empty {
+			return nil, errors.Errorf("layer %q is not unpacked", state.layer.Blob.Digest)
+		}
+	}
+	log.G(ctx).WithFields(logrus.Fields{
+		"desc":  manifestDesc,
+		"chain": chain,
+	}).Debug("image unpacked")
+	return chain, nil
+}
+
+func (u *unpacker) inform(f images.Handler) images.HandlerFunc {
+	return func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		children, err := f.Handle(ctx, desc)
+		if err != nil {
+			return children, err
+		}
+
+		switch desc.MediaType {
+		case images.MediaTypeDockerSchema1Manifest:
+			if len(children) == 0 {
+				return children, errors.Errorf("unexpected children %v", children)
+			}
+			u.manifestCh <- manifestUpdate{
+				descriptor: desc,
+				children:   children,
+			}
+		case images.MediaTypeDockerSchema2Manifest, ocispec.MediaTypeImageManifest:
+			if len(children) == 0 {
+				return children, errors.Errorf("unexpected children %v", children)
+			}
+			var descs []ocispec.Descriptor
+			for _, child := range children {
+				if child.MediaType == images.MediaTypeDockerSchema2Config ||
+					child.MediaType == ocispec.MediaTypeImageConfig {
+					continue
+				}
+				descs = append(descs, child)
+			}
+			u.manifestCh <- manifestUpdate{
+				descriptor: desc,
+				children:   descs,
+			}
+		case images.MediaTypeDockerSchema2Config, ocispec.MediaTypeImageConfig:
+			p, err := content.ReadBlob(ctx, u.c.ContentStore(), desc)
+			if err != nil {
+				return children, err
+			}
+
+			var i ocispec.Image
+			if err := json.Unmarshal(p, &i); err != nil {
+				return children, err
+			}
+
+			var diffIDs []ocispec.Descriptor
+			for _, diffID := range i.RootFS.DiffIDs {
+				diffIDs = append(diffIDs, ocispec.Descriptor{
+					MediaType: ocispec.MediaTypeImageLayer,
+					Digest:    diffID,
+				})
+			}
+			u.manifestCh <- manifestUpdate{
+				descriptor: desc,
+				children:   diffIDs,
+			}
+		default:
+			u.layerCh <- desc
+		}
+		return children, nil
+	}
+}
+
+// wait waits for the unpacker to finish and collect unpack result.
+// NOTE: wait can only be called once, and must be called after image
+// fetch finishes, either succeeds or fails.
+func (u *unpacker) wait(ctx context.Context, img Image) error {
+	close(u.manifestCh)
+	close(u.layerCh)
+
+	// unpacker only returns nil err when the image is fully unpacked.
+	var res unpackResult
+	select {
+	case res = <-u.resCh:
+		if res.err != nil {
+			return res.err
+		}
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	if img == nil {
+		return errors.New("unexpected nil image")
+	}
+
+	config, err := img.Config(ctx)
+	if err != nil {
+		return err
+	}
+
+	chainID := identity.ChainID(res.chain).String()
```

Comment:
There should be some check here that the image and chain ID are aligned. I think my preference would be that instead of just returning the chain, return a combination of chain and config, then use the config directly from there.

Reply:
I think the problem is that we don't have config for schema1 until after convert.

Reply:
Added validation.

This is bitter sweet to me. I hate the code structure, but to support schema1 I have to do it in this way. :(